wolfcrypt/src/evp.c

```diff
@@ -6803,6 +6805,870 @@ void wolfSSL_EVP_PKEY_free(WOLFSSL_EVP_PKEY* key)
+
+        XMEMCPY(buff + idx, outHex, 2);
+        idx += 2;
+
+        if (i != left - 1) {
+            wsz = 1;
+            if (idx + wsz <= EVP_PKEY_PRINT_LINE_WIDTH_MAX) {
+                XMEMSET(buff + idx, ':', wsz);
+                idx += wsz;
+            }
+        }
+    }
+
+    if (wolfSSL_BIO_write(out, buff, idx) < 0)
+        return WOLFSSL_FAILURE;
+    if (wolfSSL_BIO_write(out, "\n", 1) < 0)
+        return WOLFSSL_FAILURE;
+
+#ifdef WOLFSSL_SMALL_STACK
+    XFREE(buff, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+#endif
+    return WOLFSSL_SUCCESS;
+}
+#if !defined(NO_RSA)
+/* PrintPubKeyRSA is a helper function for wolfSSL_EVP_PKEY_print_public
+ * to parse a DER format RSA public key specified in the second parameter.
+ * Returns 1 on success, 0 on failure.
+*/
+static int PrintPubKeyRSA(WOLFSSL_BIO* out, const byte* pkey, int pkeySz,
+    int indent, int bitlen, ASN1_PCTX* pctx)
+{
+
+    byte   buff[8] = { 0 };
+    word32 inOutIdx = 0;
+    int length = 0;
+    word32 nSz;             /* size of modulus */
+    word32 eSz;             /* size of public exponent */
+    byte*  n   = NULL;
+    byte*  e   = NULL;      /* pointer to modulus/exponent */
+    word32 localIdx;
+    word32 oid;
+    byte   tag;
+    int idx;
+    int wsz;
+    word32 i;
+    word32 exponent = 0;
+    word32 outSz;
+    byte   outHex[3];
+    mp_int a;
+
+    (void)pctx;
+
+    if( mp_init(&a) != 0)
+        return WOLFSSL_FAILURE;
+
+    if (indent < 0)
+        indent = 0;
+    if (indent > EVP_PKEY_PRINT_INDENT_MAX)
+        indent = EVP_PKEY_PRINT_INDENT_MAX;
+
+    if (GetSequence(pkey, &inOutIdx, &length, pkeySz) < 0)
+        return WOLFSSL_FAILURE;
+
+    localIdx = inOutIdx;
+    if (GetASNTag(pkey, &localIdx, &tag, pkeySz) < 0)
+        return WOLFSSL_FAILURE;
+
+    if (tag != ASN_INTEGER) {
+        /* not from decoded cert, will have algo id, skip past */
+        if (GetSequence(pkey, &inOutIdx, &length, pkeySz) < 0)
+            return WOLFSSL_FAILURE;
+
+        if (GetObjectId(pkey, &inOutIdx, &oid, oidIgnoreType, pkeySz) != 0)
+            return WOLFSSL_FAILURE;
+
+        /* Option NULL ASN.1 tag */
+        if (inOutIdx  >= (word32)pkeySz)
+            return WOLFSSL_FAILURE;
+
+        localIdx = inOutIdx;
+        if (GetASNTag(pkey, &inOutIdx, &tag, pkeySz) < 0)
+            return WOLFSSL_FAILURE;
+
+        if (tag != ASN_TAG_NULL)
+            return WOLFSSL_FAILURE;
+
+        inOutIdx ++;
+
+        /* should have bit tag length and seq next */
+        if (CheckBitString(pkey, &inOutIdx, NULL, pkeySz, 1, NULL) != 0)
+            return WOLFSSL_FAILURE;
+
+        if (GetSequence(pkey, &inOutIdx, &length, pkeySz) < 0)
+            return WOLFSSL_FAILURE;
+    }
+    /* Get modulus */
+    if (GetASNTag(pkey, &inOutIdx, &tag, pkeySz ) < 0)
+        return WOLFSSL_FAILURE;
+
+    if (tag != ASN_INTEGER)
+        return WOLFSSL_FAILURE;
+
+    if (GetLength(pkey, &inOutIdx, &length, pkeySz) < 0)
```

Comment:
length can be 0, in which case the read of

```
&pkey[inOutIdx]
```

will be out of bounds. Can we add either a <= check of 0 here or a check on inOutIdx before it is read from pkey. Both here and with the variable 'e'.

Reply:
Added checks of 0.